File: 200-number-of-islands/number-of-islands.cpp
class Solution {
private:
    void bfs(int r, int c, vector<vector<int>>&visited, int m, int n, vector<vector<char>>& grid){
        queue<pair<int,int>>q;
        q.push({r,c});
        visited[r][c]=1;
        int dr[]={-1,0,+1,0};
        int dc[]={0,+1,0,-1};
        while(!q.empty()){
            int row=q.front().first;
            int col=q.front().second;
            q.pop();
            for(int i=0;i<4;i++){
                int newr=row+dr[i];
                int newc=col+dc[i];

                if(newr>=0 && newr<m && newc>=0 && newc<n && visited[newr][newc]!=1 && grid[newr][newc]=='1'){
                    q.push({newr,newc});
                    grid[newr][newc]='0';
                }
            }
        }
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int island=0;
        int m=grid.size();
        int n=grid[0].size();
        vector<vector<int>>visited(m,vector<int>(n,0));
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]=='1'){
                    island+=1;
                    bfs(i,j,visited,m,n,grid);
                }
            }
        }
        return island;
    }
};

File: 200-number-of-islands/number-of-islands.py
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        vis = [[0] * n for _ in range(m)]

        def bfs(r, c):
            q = deque()
            q.append((r, c))
            vis[r][c] = 1

            directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

            while q:
                row, col = q.popleft()
                for dr, dc in directions:
                    nr, nc = row + dr, col + dc
                    if 0 <= nr < m and 0 <= nc < n and vis[nr][nc] == 0 and grid[nr][nc] == "1":
                        vis[nr][nc] = 1
                        q.append((nr, nc))

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and vis[i][j] == 0:
                    bfs(i, j)
                    count += 1

        return count
